Add end-to-end test app

NSeydoux opened the pull request:
This adds a simple end-to-end test app allowing the user to log in, connect a websocket and see the events stream. The app should be linked to the local code so that the built current library is under tests.

Note that the testcafé tests are left unchanged, and simply moved to a `test` subfolder of `e2e/browser`, sibling of the newly created app's `src` directory. using this app in playwright end-to-end browser-based tests will be done in an upcoming PR.

NSeydoux committed:
Add end-to-end test app

This adds a simple end-to-end test app allowing the user to log in, connect a websocket and see the events stream. The app should be linked to the local code so that the built current library is under tests.

NSeydoux committed:
Fix lint issues

ThisIsMissEm commented on e2e/browser/src/components/notifications/index.tsx:
```diff
@@ -0,0 +1,149 @@
+  const [messageBus, setMessageBus] = useState<any[]>([]);
+
+  useEffect(() => {
+    if (getDefaultSession().info.webId !== undefined) {
+      getPodUrlAll(getDefaultSession().info.webId as string, {
+        fetch: getDefaultSession().fetch,
```
```suggestion
    const session = getDefaultSession();
    if (session.info.webId !== undefined) {
      getPodUrlAll(session.info.webId as string, {
        fetch: session.fetch,
```

NSeydoux replied:
4c7998b2187b6ff50440e31f8eb5d736a9621d6d

ThisIsMissEm commented on e2e/browser/src/.eslintrc.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "next/core-web-vitals"
```
might be good to use our rules here instead

ThisIsMissEm commented on e2e/browser/src/components/appContainer/index.tsx:
```diff
@@ -0,0 +1,88 @@
+} from "@inrupt/solid-client-authn-browser";
+import Notifications from "../notifications/";
+
+const REDIRECT_URL = "http://localhost:3000/";
```
probably better to just use `window.location.href` here

NSeydoux replied:
c3d761f15ee49e1d0154324348f60167883d2d65

ThisIsMissEm commented on e2e/browser/src/components/appContainer/index.tsx:
```diff
@@ -0,0 +1,88 @@
+
+const REDIRECT_URL = "http://localhost:3000/";
+const APP_NAME = "Notifications browser-based tests app";
+const issuer = "https://login.inrupt.com/";
```
ideally this should come from an environment variable, along with docs as to why it's this specific environment

NSeydoux replied:
There are some changes to this intent in #232, I'll add this there.

NSeydoux replied:
Actually the issuer does come from the environment (it is set by playwright), the app name is not environment-specific, and once using `window.location.href` neither is the redirect URL.

ThisIsMissEm replied:
Yup, you just need to call this `DEFAULT_ISSUER` instead of `issuer` and then you can use it as the initial value in your `useState` for the issuer

NSeydoux replied:
Done in https://github.com/inrupt/solid-client-notifications-js/pull/229/commits/aa86165ebd61620ee42d0c6cdf9a27538606b1fe :)

ThisIsMissEm commented on e2e/browser/src/components/appContainer/index.tsx:
```diff
@@ -0,0 +1,88 @@
+        restorePreviousSession: true,
+      });
+      if (info !== undefined) {
+        setLoggedIn(info.isLoggedIn);
```
You may actually want to store the entire info object, and then pass that down to child components

NSeydoux replied:
f7acd4ff3913da3214e603aa36015fa3a7f1b138

ThisIsMissEm commented on e2e/browser/src/components/appContainer/index.tsx:
```diff
@@ -0,0 +1,88 @@
+  const handleLogin = async () => {
+    // Login will redirect the user away so that they can log in the OIDC issuer,
+    // and back to the provided redirect URL (which should be controlled by your app).
+    await login({
```
I'd suggest wrapping in a try/catch, and then logging the error if there is one.

ThisIsMissEm commented on e2e/browser/src/components/appContainer/index.tsx:
```diff
@@ -0,0 +1,88 @@
+
+export default function AppContainer() {
+  const [loggedIn, setLoggedIn] = useState<boolean>(false);
+  const [issuer, setIssuer] = useState<string>("https://login.inrupt.com");
```
should reuse `issuer` from line 12, but rename that variable `DEFAULT_ISSUER`

NSeydoux replied:
aa86165ebd61620ee42d0c6cdf9a27538606b1fe

ThisIsMissEm commented on e2e/browser/src/components/notifications/index.tsx:
```diff
@@ -0,0 +1,149 @@
+      socket.on("error", () => setConnectionStatus("error"));
+      socket.on("message", (message) => {
+        console.debug("New message", { message });
+        messageBus.push(JSON.parse(message));
+        // Reverse the message bus so that the latest message appears first
+        setMessageBus([...messageBus.reverse()]);
```
this is odd, you'd be better off doing:
```suggestion
        const parsedMessage = JSON.parse(message);
        setMessageBus([parsedMessage, ...messageBus]);
```

NSeydoux replied:
That's what I initially did, but for some reason I don't understand, written like this the message array only ever contains one element, the latest message.

ThisIsMissEm replied:
hmm, maybe:
```
setMessageBus((previousMessageBus) => [parsedMessage, ...previousMessageBus]);
```

per https://reactjs.org/docs/hooks-reference.html#functional-updates

NSeydoux replied:
Aha, that does the trick, thanks ! 246e4557b44aff30affeca6d04e052b76f564a9b

ThisIsMissEm reviewed (commented):
A few changes suggested, and I'd also encourage you to call the `src` directory `testapp`

NSeydoux committed:
Store getDefaultSession in a variable

NSeydoux committed:
Use current location instead of default

NSeydoux committed:
Consistently use default issuer

NSeydoux committed:
Pass session info to child components

NSeydoux committed:
Merge useEffect

NSeydoux committed:
Rename test app folder

NSeydoux commented on e2e/browser/src/components/notifications/index.tsx:
```diff
@@ -0,0 +1,149 @@
+  }, [parentContainerUrl]);
+
+  useEffect(() => {
+    if (socket !== undefined) {
```
aa86165ebd61620ee42d0c6cdf9a27538606b1fe

NSeydoux committed:
Fix linting issues

NSeydoux committed:
Use functional state update

To update a component's state using its previous state, a functional update should be used.
